Examples/CaseStudies/SwiftUICaseStudies/02-Effects-WebSocket.swift

```diff
@@ -0,0 +1,374 @@
+      .cancellable(id: WebSocketId())
+  }
+
+  switch action {
+  case .alertDismissed:
+    state.alert = nil
+    return .none
+
+  case .connectButtonTapped:
+    switch state.connectivityState {
+    case .connected, .connecting:
+      state.connectivityState = .disconnected
+      return .cancel(id: WebSocketId())
+
+    case .disconnected:
+      state.connectivityState = .connecting
+      return environment.webSocket.open(WebSocketId(), URL(string: "wss://echo.websocket.org")!, [])
+        .receive(on: environment.mainQueue)
+        .map(WebSocketAction.webSocket)
+        .eraseToEffect()
+        .cancellable(id: WebSocketId())
+    }
+
+  case let .messageToSendChanged(message):
+    state.messageToSend = message
+    return .none
+
+  case .pingResponse:
+    // Ping the socket again in 10 seconds
+    return sendPingEffect
+
+  case let .receivedSocketMessage(.success(.string(string))):
+    state.receivedMessages.append(string)
+
+    // Immediately ask for the next socket message
+    return receiveSocketMessageEffect
+
+  case .receivedSocketMessage(.success):
+    // Immediately ask for the next socket message
+    return receiveSocketMessageEffect
+
+  case .receivedSocketMessage(.failure):
+    return .none
+
+  case .sendButtonTapped:
+    let messageToSend = state.messageToSend
+    state.messageToSend = ""
+
+    return environment.webSocket.send(WebSocketId(), .string(messageToSend))
+      .eraseToEffect()
+      .map(WebSocketAction.sendResponse)
+
+  case let .sendResponse(error):
+    if error != nil {
+      state.alert = "Could not send socket message. Try again."
+    }
+    return .none
+
+  case let .webSocket(.didClose(code, _)):
+    state.connectivityState = .disconnected
+    return .cancel(id: WebSocketId())
+
+  case let .webSocket(.didBecomeInvalidWithError(error)),
+    let .webSocket(.didCompleteWithError(error)):
+    state.connectivityState = .disconnected
+    if error != nil {
+      state.alert = "Disconnected from socket for some reason. Try again."
+    }
+    return .cancel(id: WebSocketId())
+
+  case .webSocket(.didOpenWithProtocol):
+    state.connectivityState = .connected
+    return .merge(
+      receiveSocketMessageEffect,
+      sendPingEffect
+    )
+  }
+}
+.debug()
+
+struct WebSocketView: View {
+  let store: Store<WebSocketState, WebSocketAction>
+
+  var body: some View {
+    WithViewStore(self.store) { viewStore in
+      VStack(alignment: .leading) {
+        Text(template: readMe, .body)
+          .padding([.bottom])
+
+        HStack {
+          TextField(
+            "Message to send",
+            text: viewStore.binding(
+              get: { $0.messageToSend }, send: WebSocketAction.messageToSendChanged)
+          )
+
+          Button(
+            viewStore.connectivityState == .connected
+              ? "Disconnect"
+              : viewStore.connectivityState == .disconnected
+                ? "Connect"
+                : "Connecting..."
```

Comment:
I usually find this style of ternary a lil easier to read:
```suggestion
            viewStore.connectivityState == .connected ? "Disconnect"
              : viewStore.connectivityState == .disconnected ? "Connect"
              : "Connecting..."
```